### PR TITLE
[dtensor] fast return aten.detach

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112931
* #112927

This PR makes a fast return when dispatching to aten.detach call. The
reason is that aten.detach is a call relates to autograd but wrapper
subclass usually deal with autograd history on the wrapper directly, not
the underlying shard, also torch_dispatch won't have any info about
autograd anyways.

So as long as we make sure wrapper subclass interact with underlying
tensor correctly (i.e. via from_local, view_as), it should work

### CODE REVIEW

File: torch/distributed/_tensor/dispatch.py [hunk 1]
```diff
@@ -130,6 +130,13 @@ def _operator_dispatch(
         assert isinstance(args[0], torch.Tensor)
         assert isinstance(args[1], torch.Tensor)
         return args[0].shape == args[1].shape, None, None  # type: ignore[return-value]
+    elif op_call == aten.detach.default:
+        # directly wrap a new DTensor with the input DTensor's local_shard and spec, this
```
Comment (wconstab) [inline torch/distributed/_tensor/dispatch.py:134]: can you convince me this is true?

in particular, i was wondering why we don't have to actually call .detach() on the _local_tensor before we wrap it.

Comment (wanchaol) [reply torch/distributed/_tensor/dispatch.py:134]: Sure, so here is my understanding about how detach works:
1. https://pytorch.org/docs/stable/generated/torch.Tensor.detach.html#torch.Tensor.detach detach basically creates a new tensor that holds the same underlying data for autograd purposes, i.e it tells autograd to create a new graph that is detached from the existing graph.
2. in __torch_dispatch__, we already lowered to the dispatching system that's BELOW autograd, so any autograd related information already gets away, the only thing we need to make sure inside `__torch_dispatch__` is to creates a new wrapper subclass for autograd to consume later. That's the reason that local_shard don't need to call detach
3. so as long as user don't try to work with the `_local_tensor` directly, our semantics would be correct. Because whenever we create/to from torch.tensor, we always makes sure the autograd history get perserved (i.e. in both `from_local/to_local` we have `view_as` call to create new objects, and make sure local_shard autograd and wrapper autograd interacts correctly, we don't have concerns like inplace autograd metadata mutations.

Comment (wconstab) [reply torch/distributed/_tensor/dispatch.py:134]: so could the _local_tensor already be a part of another autograd graph (from before we wrapped it with DTensor)?  I guess, that may be OK and we can leave it alone, becauae the user only tried to 'detach' the DTensor which may exist in a different autograd graph?

Comment (albanD) [reply torch/distributed/_tensor/dispatch.py:134]: > so could the _local_tensor already be a part of another autograd graph (from before we wrapped it with DTensor)?

It is not forbidden by the subclass system. But for DTensor, that would be a logical error: the DTensor is the one working with autograd, the _local_tensor never does. In particular, as described in point 3 above, inside from/to_local, we are careful to make sure that the plain Tensor is not the same as the one stored as _local_tensor (using view_as in no_grad mode for example).

**Review by fduwjj — APPROVED** [review]

**Review by XilunWu — APPROVED** [review]
Good to learn that handling `detach()` in DTensor is simple and clean.
